Utilities/Thread.cpp

```diff
@@ -2019,20 +2019,100 @@ u16 thread_ctrl::get_affinity_mask(thread_class group)
 		}
 		case native_core_arrangement::amd_ccx:
 		{
-			u16 spu_mask, ppu_mask, rsx_mask;
-			if (thread_count >= 16)
+			u32 spu_mask, ppu_mask, rsx_mask;
+			const auto system_id = utils::get_system_info();
+			if (thread_count >= 32)
 			{
-				// Threadripper, R7
-				// Assign threads 8-16
-				// It appears some windows code is bound to lower core addresses, binding 8-16 is alot faster than 0-7
-				ppu_mask = spu_mask = 0b1111111100000000;
-				rsx_mask = all_cores_mask;
+				if (system_id.find("3950X") != std::string::npos)
+				{
+					// zen2
+					// Ryzen 9 3950X
+					// Assign threads 9-30
+					ppu_mask = 0b001111000000000000000000000000;
+					spu_mask = 0b000000111111110000000000000000;
+					rsx_mask = 0b000000000000000011111100000000;
+				}
+				else if (system_id.find("2970WX") != std::string::npos)
+				{
+					// zen+
+					// Threadripper 2970WX
+					// Assign threads 9-24
+					ppu_mask = 0b000000111111000000000000;
+					spu_mask = ppu_mask;
+					rsx_mask = 0b111111000000000000000000;
+				}
+				else
+				{
+					// zen(+)
+					// Threadripper 1950X/2950X/2990WX
+					// Assign threads 17-30
+					ppu_mask = 0b00000000111111110000000000000000;
+					spu_mask = ppu_mask;
+					rsx_mask = 0b00111111000000000000000000000000;
+				}
+
+			}
+			else if (thread_count == 24)
+			{
+				if (system_id.find("3900X") != std::string::npos)
+				{
+					// zen2
+					// Ryzen 9 3900X
+					// Assign threads 7-22
+					ppu_mask = 0b001111000000000000000000;
+					spu_mask = 0b000000111111000000000000;
+					rsx_mask = 0b000000000000111111000000;
+				}
+				else
+				{
+					// zen(+)
+					// Threadripper 1920X/2920X
+					// Assign threads 13-24
+					ppu_mask = 0b000000111111000000000000;
+					spu_mask = ppu_mask;
+					rsx_mask = 0b111111000000000000000000;
+				}
+			}
+			else if (thread_count == 16)
+			{
+				if (system_id.find("3700X") != std::string::npos || system_id.find("3800X") != std::string::npos)
+				{
+					// Ryzen 7 3700/3800 (x)
+					// Assign threads 1-16
+					ppu_mask = 0b0000000011110000;
+					spu_mask = 0b1111111100000000;
+					rsx_mask = 0b0000000000001111;
+				}
+				else
+				{
+					// zen(+)
+					// Ryzen 7, Threadripper 
+					// Assign threads 3-16
+					ppu_mask = 0b1111111100000000;
+					spu_mask = ppu_mask;
+					rsx_mask = 0b0000000000111100;
```

Comment:
Do not restrict RSX. Its a heavy set that does benefit from being scheduled on same block as SPU which is why I intentionally left it as all cores on gen1 ryzen. Making it separate was actually quite a bit slower.

Reply:
Is more speed like that, set rsx on all core mask is not very good, like that improve 1 2 fps on zen1 and very more on zen2.
And for information all core mask using all time First thread by orderso just set rsx on all core mask get few first thread using, is for that get more speed and stability like that.